Wkcuber with typer (#903)
scalableminds/webknossos-libs: wkcuber/README.md

# WEBKNOSSOS cuber (wkcuber)
[![PyPI version](https://img.shields.io/pypi/v/wkcuber)](https://pypi.python.org/pypi/wkcuber)

> [![Supported Python Versions](https://img.shields.io/pypi/pyversions/wkcuber.svg)](https://pypi.python.org/pypi/wkcuber)

Reviewer: I would keep the little icons/shields and update them to wk-libs or copy from that readme

## Features

wkcuber offers some commands to work with WEBKNOSSOS datasets:

> - `wkcuber compress`: Compress a WEBKNOSSOS dataset

Reviewer: @normanrz  More general question: Do we want to keep the `wkcuber` name for the command or switch to just `webknossos <command>`?

> - `wkcuber convert`: Convert a image stack to a WEBKNOSSOS dataset

Reviewer: ```suggestion
- `wkcuber convert`: Convert an image stack (e.g., `tiff`, `jpg`, `png`, `bmp`, `dm3`, `dm4`) to a WEBKNOSSOS dataset
```

> - `wkcuber download`: Download a WEBKNOSSOS dataset from a remote location

Reviewer: ```suggestion
- `wkcuber download`: Download a dataset from a WEBKNOSSOS server as WKW format
```

- `wkcuber downsample`: Downsample a WEBKNOSSOS dataset
- `wkcuber upload`: Upload a local WEBKNOSSOS dataset to a remote location
- `wkcuber upsample`: Upsample a WEBKNOSSOS dataset

## Supported input formats

- Standard image formats, e.g. `tiff`, `jpg`, `png`, `bmp`
- Proprietary image formats, e.g. `dm3`

> - Raw binary files

Reviewer: I assume even more biology image formats are possible through the `bioformats` library. Perhaps, that require installing additional dependencies. See `--all-deps`(?) install option in the wk-libs. If that works we should list that here as it would double and triple the amount of supported image file formats. Please check.

Author: I updated the installation instructions to install webknossos[all] to get support for most image types.

- `wkcuber download`: Download a WEBKNOSSOS dataset from a remote location
- `wkcuber downsample`: Downsample a WEBKNOSSOS dataset
- `wkcuber upload`: Upload a local WEBKNOSSOS dataset to a remote location
- `wkcuber upsample`: Upsample a WEBKNOSSOS dataset

## Supported input formats

* Standard image formats, e.g. `tiff`, `jpg`, `png`, `bmp`
* Proprietary image formats, e.g. `dm3`
* Tiled image stacks (used for Catmaid)
* KNOSSOS cubes
* NIFTI files
* Raw binary files
- Standard image formats, e.g. `tiff`, `jpg`, `png`, `bmp`
- Proprietary image formats, e.g. `dm3`
- Raw binary files

## Installation
### Python 3 with pip from PyPi
- `wkcuber` requires at least Python 3.8

```bash
# Make sure to have lz4 installed:
# Mac: brew install lz4
# Ubuntu/Debian: apt-get install liblz4-1
# CentOS/RHEL: yum install lz4

pip install wkcuber
```

### Docker
Use the CI-built image: [scalableminds/webknossos-cuber](https://hub.docker.com/r/scalableminds/webknossos-cuber/). Example usage `docker run -v <host path>:/data --rm scalableminds/webknossos-cuber wkcuber --layer_name color --scale 11.24,11.24,25 --name great_dataset /data/source/color /data/target`.

### Python 3 with pip from PyPi

> ## Usage

Reviewer 2: The CI tests Python 3.8+

- `wkcuber` requires at least Python 3.10

```bash
# Convert arbitrary, supported input files into wkw datasets. This sets reasonable defaults, but see other commands for customization.

> python -m wkcuber \

Reviewer 2: Some instructions/examples for how to use the wkcuber are necessary in the readme.

Author: I added some examples in the docs.

# Compress data in-place (mostly useful for segmentation)
python -m wkcuber.compress --layer_name segmentation data/target

# Compress data copy (mostly useful for segmentation)
python -m wkcuber.compress --layer_name segmentation data/target data/target_compress

# Create metadata
python -m wkcuber.metadata --name great_dataset --scale 11.24,11.24,25 data/target

# Refresh metadata so that new layers and/or magnifications are picked up
python -m wkcuber.metadata --refresh data/target

# Recubing an existing dataset
python -m wkcuber.recubing --layer_name color --dtype uint8 /data/source/wkw /data/target
pip install wkcuber

# Check two datasets for equality
python -m wkcuber.check_equality /data/source /data/target
# to install auto completion as well use:
wkcuber --install-completion
```

### Parallelization

Most tasks can be configured to be executed in a parallelized manner. Via `--distribution_strategy` you can pass `multiprocessing`, `slurm` or `kubernetes`. The first can be further configured with `--jobs` and the latter via `--job_resources='{"mem": "10M"}'`. Use `--help` to get more information.

> ### Zarr support

Reviewer 2: Can you check that the Zarr support still works? I think there needs to be some supporting code in wkcuber to translate the env vars into UPaths.

# WEBKNOSSOS cuber (wkcuber)
[![PyPI version](https://img.shields.io/pypi/v/wkcuber)](https://pypi.python.org/pypi/wkcuber)
[![Supported Python Versions](https://img.shields.io/pypi/pyversions/wkcuber.svg)](https://pypi.python.org/pypi/wkcuber)

Python CLI for creating and working with [WEBKNOSSOS](https://webknossos.org/) [WKW](https://github.com/scalableminds/webknossos-wrap) datasets. WKW is a container format for efficiently storing large, scale 3D image data as found in (electron) microscopy.

## Features

> wkcuber offers some commands to work with WEBKNOSSOS datasets:

Reviewer: ```suggestion
wkcuber offers many useful commands to work with WEBKNOSSOS datasets:
```

- `wkcuber convert`: Convert an image stack (e.g., `tiff`, `jpg`, `png`, `bmp`, `dm3`, `dm4`) to a WEBKNOSSOS dataset
- `wkcuber download`: Download a dataset from a WEBKNOSSOS server as WKW format
- `wkcuber downsample`: Downsample a WEBKNOSSOS dataset
- `wkcuber upload`: Upload a local WEBKNOSSOS dataset to a remote location
- `wkcuber upsample`: Upsample a WEBKNOSSOS dataset

## Supported input formats

- Standard image formats, e.g. `tiff`, `jpg`, `png`, `bmp`
- Proprietary image formats, e.g. `dm3`
- Raw binary files

## Installation

### Python 3 with pip from PyPi

- `wkcuber` requires at least Python 3.8

```bash
pip install wkcuber

# to install auto completion as well use:
wkcuber --install-completion
```

## Usage

```bash
# Convert image stacks into wkw datasets
wkcuber convert \
  --voxel-size 11.24,11.24,25 \
  --name great_dataset \
  data/source data/target

# Create downsampled magnifications
wkcuber downsample data/target
wkcuber downsample --layer_name color data/target

# Compress data in-place (mostly useful for segmentation)
wkcuber compress --layer_name segmentation data/target
wkcuber compress data/target

> # Old modules that are still usable until there is a typer replacement

Reviewer 2: ```suggestion
# These legacy modules that are still usable, but will be replaced in a future release
```

poetry run pytest tests
```

Please, run the extended test suite:
```bash

> tests/scripts/all_tests.sh

Reviewer: What about these tests? Were they removed? The `all_tests.sh` file is not part of the PR and remains unchanged.

```

PyPi releases are automatically pushed when creating a new Git tag/Github release. 

## API documentation

> Check out the [latest version of the API documentation](https://docs.webknossos.org/api/wkcuber.html).

Reviewer: I guess we publish the API documentation at https://docs.webknossos.org/api/wkcuber.html

@normanrz I am not sure about the usefulness of the API documents and if we still need them as the wkcuber is now such a lightweight wrapper. 

@markbader If we indeed remove them, then please remove `docs/generate.sh` and  adjust the CI and general documentation generation workflows accordingly.

Author: The webknossos cli is now described in the webknossos-py docs and the CI is adapted accordingly.

### Generate the API documentation
Run `docs/generate.sh` to open a server displaying the API docs. `docs/generate.sh --persist` persists the html to `docs/api`.

> ## Test Data Credits

Reviewer 2: I guess, we still use this testdata.

Author: Test data credits are now from webknossos readme.

tests/scripts/all_tests.sh
```

### Generate the API documentation
Run `docs/generate.sh` to open a server displaying the API docs. `docs/generate.sh --persist` persists the html to `docs/api`.

## License
AGPLv3

> Copyright scalable minds

Reviewer 2: ```suggestion
Copyright scalable minds

```
